zilliqa/src/db.rs

```diff
@@ -221,6 +235,111 @@ impl Db {
+        parent.verify_hash()?;
+
+        if block.parent_hash() != parent.hash() {
+            return Err(anyhow!("Invalid checkpoint file: parent's blockhash does not correspond to checkpoint block"));
+        }
+
+        let shard_number: u64 = lines
+            .next()
+            .ok_or(anyhow!(
+                "Invalid checkpoint file: missing shard id on line 3"
+            ))??
+            .parse()?;
+        if shard_number != our_shard_id {
+            return Err(anyhow!("Invalid snapshot: chain ID mismatch. Snapshot ID: {shard_number}, our chain_id: {our_shard_id}"));
+        }
+
+        let our_max_view = self.get_latest_finalized_view()?.unwrap_or(0);
+        if block.view() <= our_max_view {
+            return Err(anyhow!("Snapshow is lower than our finalized view: snapshot is at {}, we are already at {}", block.view(), our_max_view));
+        }
+
+        // then decode state
+        for (idx, line) in lines.enumerate() {
+            let idx = idx + 2; // +1 because first line is just serialised block, +1 for 1-indexing
+            let line = line?;
+            let (account, trie) = line
+                .split_once(';')
+                .ok_or(anyhow!("Invalid checkpoint file at line {idx}"))?;
+            let (account_hash, serialized_account) = account.split_once(':').ok_or(anyhow!(
+                "Invalid checkpoint file: invalid state account information at line {idx}"
+            ))?;
+            let serialized_account = hex::decode(serialized_account)?;
+            let mut account_trie = EthTrie::new(trie_storage.clone());
+            for (storage_idx, storage_entry) in trie.split(',').enumerate() {
+                if storage_entry.is_empty() {
+                    continue;
+                }
+                let (key, val) = storage_entry.split_once(':').ok_or(anyhow!(
+                    "Invalid checkpoint file: invalid storage entry at line {idx}, index {storage_idx}",
+                ))?;
+                account_trie.insert(&hex::decode(key)?, &hex::decode(val)?)?;
+            }
+            let account_trie_root =
+                bincode::deserialize::<Account>(&serialized_account)?.storage_root;
+            if account_trie.root_hash()?.as_slice() != account_trie_root {
+                return Err(anyhow!(
+                    "Invalid checkpoint file: account trie root hash mismatch, at line {idx}: calculated {}, checkpoint file contained {}", hex::encode(account_trie.root_hash()?.as_slice()), hex::encode(account_trie_root)
+                ));
+            }
+            state_trie.insert(&hex::decode(account_hash)?, &serialized_account)?;
+        }
+        if state_trie.root_hash()? != block.state_root_hash().0 {
+            return Err(anyhow!("Invalid checkpoint file: state root hash mismatch"));
+        }
+
+        let block_ref = &block; // for moving into the closure
+        self.with_sqlite_tx(move |tx| {
```

Comment:
Good point.
Now I _think_ it's not actually that bad because you'll fetch blocks from the network, realise you're out of date, and request a sync - and then you'll just magically find that the blocks your receive are in your database, so you update your "latest view", request the next batch, magically find you already had those blocks too, etc...
But I'm not certain this is _guaranteed_ to work without any weirdness or misbehaviour, so agree that it would be safer to avoid this scenario entirely.